src/lambda_vm/interpreter.ml

```diff
@@ -0,0 +1,192 @@
+    match args with
+    | [] -> V_primitive { args = [arg]; prim }
+    | [left] -> f left arg
+    | _ -> raise Over_applied_primitives in
+
+  let op2_shift f =
+    let f left right = f left (Int64.to_int right) in
+    op2 f in
+  match prim with
+  | P_neg -> op1 Int64.neg
+  | P_add -> op2 Int64.add
+  | P_sub -> op2 Int64.sub
+  | P_mul -> op2 Int64.mul
+  | P_div -> op2 Int64.div
+  | P_rem -> op2 Int64.rem
+  | P_and -> op2 Int64.logand
+  | P_or -> op2 Int64.logor
+  | P_xor -> op2 Int64.logxor
+  | P_lsl -> op2_shift Int64.shift_left
+  | P_lsr -> op2_shift Int64.shift_right_logical
+  | P_asr -> op2_shift Int64.shift_right
+
+(* TODO: gas must be ref *)
+let rec eval ~stack gas env code =
+  let eval_call env code = eval ~stack:(stack - 1) gas env code in
+  let eval_jump env code = eval ~stack gas env code in
+
+  check_stack ~stack;
+  burn_gas gas env code;
+
+  match code with
+  | E_var var -> (
+    (* TODO: gas cost for this *)
+    (* TODO: should this be a hashtbl? Or an array? *)
+    match Env.find var env with
+    | Some value -> value
+    | None ->
+      (* TODO: could we eliminate this using GADTs? *) raise Undefined_variable)
+  | E_lam (param, body) -> V_closure { env; param; body }
+  | E_app { funct; arg } -> (
+    let funct = eval_call env funct in
+    let arg = eval_call env arg in
+    match funct with
+    | V_pair _
+    | V_int64 _ ->
+      raise Value_is_not_function
+    | V_closure { env; param; body } ->
+      let env = Env.add param arg env in
+      eval_jump env body
+    | V_primitive { args; prim } -> eval_prim prim ~arg ~args)
+  | E_const value -> V_int64 value
+  | E_prim prim -> V_primitive { args = []; prim }
+  | E_if { predicate; consequent; alternative } -> (
+    let predicate = eval_call env predicate in
+    match predicate with
+    | V_int64 0L -> eval_jump env alternative
+    | V_int64 _ -> eval_jump env consequent
+    | V_pair _
+    | V_closure _
+    | V_primitive _ ->
+      raise Value_is_not_int64)
+  | E_pair { first; second } ->
+    let first = eval_call env first in
+    let second = eval_call env second in
+    V_pair { first; second }
+  | E_fst pair -> (
+    let pair = eval_call env pair in
+    match pair with
+    | V_pair { first; second = _ } -> first
+    | V_int64 _
+    | V_closure _
+    | V_primitive _ ->
+      raise Value_is_not_pair)
+  | E_snd pair -> (
+    let pair = eval_call env pair in
+    match pair with
+    | V_pair { first = _; second } -> second
+    | V_closure _
+    | V_int64 _
+    | V_primitive _ ->
+      raise Value_is_not_pair)
+let eval gas env code =
+  let stack = max_stack_depth in
+  eval ~stack gas env code
+
+type script_result = {
+  storage : Ir.value;
+  operations : unit;
+}
+
+let execute gas ~arg script =
+  let { param; code } = script in
+  let env = Env.add param arg Env.empty in
+  let output = eval gas env code in
+  let storage, (operations, ()) = Pattern.(parse output script_result) in
+  { storage; operations }
```

Comment:
If `storage` is the output of the execution, and it is a `Ir.value` then `Ir.value` should be exposed and accessible externally so the client code can interpret the results. I also think that the results shouldn't be bound to a specific pattern on the output of the script, one could have different ways of running it (Tezos has offchain views for example).

I like the idea of validating the output against a specific pattern, but that would be better if the user could define it, but that's just my point of view.

Reply:
Then how would we know where is the operations? The result is the storage to be stored and the operations to be triggered, to do that it needs to be an specific pattern, Tezos also does that.